### ISSUE TITLE
How the Certified Radius was Calculated in SplitMethodDerandomized

### ISSUE THREAD

**Issue by Rui-Wang-813** [alevine0/smoothingSplittingNoise#1]
Hi! I am wondering how this certified radius is calculated in noises.py, class SplitMethodDerandomized. Code is:

```
def classify_and_certify_l1_exact_from_counts(self, counts):
        predinctionsnp = counts.cpu().numpy()
        idxsort = np.argsort(-predinctionsnp,axis=1,kind='stable')
        # valsort = -np.sort(-predinctionsnp,axis=1,kind='stable')
        valsort = np.take_along_axis(predinctionsnp, idxsort, 1)
        val =  valsort[:,0]
        idx = idxsort[:,0]
        valsecond =  valsort[:,1]
        idxsecond =  idxsort[:,1] 
        diffs = ((val - valsecond - (idxsecond <= idx))/2).astype(int)
        import pdb; pdb.set_trace()
        certs = torch.tensor(diffs).to(self.device)/255.
        torchidx = torch.tensor(idx).to(self.device)
        return torchidx, certs
```

I've read the paper but did not find specification on how the radius rho was computed. More importantly, I do not understand why there is an (idxsecond <= idx) in the diffs computation, I thought it would not matter as they are simply the label of different classes.

**Comment by alevine0** [comment]
Hi Rui,
The radius calculation is a direct consequence of the Lipschitz continuity, as explained in the introduction of the paper. Specifically, right before Equation 1:
"In particular, consider a classifier with log- its {p_A,p_B,p_C,...}, all of which are c-Lipschitz. Suppose for an input x, we have p_A(x) > p_B(x) ≥ p_C(x) ≥ ... . Also suppose the gap between the largest and the second largest logits is d (i.e. p_A(x) − p_B(x) = d). The Lipschitzness implies that for all x′ such that ||x − x′|| < d/(2c), p_A (x′ ) will still be the largest logit."

Now, in the case of deterministic smoothing, the probabilities are simply averages of the results of  exactly q * 2 * lambda discrete "hard" classifications of noisy samples, and p_A is simply the fraction of these samples for which the base classifier returns the most frequent classification result (with p_B being the fraction of these samples where the classifier returns the second-most-frequent result.) In the code, we compute the gap in the _number_ of samples classified as A versus as B (that is, "val - valsecond"); let's call this

d' := (p_A(x) - p_B(x)) * (num_samples) == d * q* 2* lambda (== "val - valsecond")


Note that this guarantees 1/(2 * lambda)-Lipschitzness, so by Equation 1, p_A(x') will remain the largest logit if ||x − x′|| < d * 2 * lambda/(2). In terms of d', this means that p_A(x') will remain the largest logit if
||x − x′|| < (d'/q * 2 * lambda) * 2 * lambda/(2)
or
||x − x′|| < d'/(q * 2)
Note that q, the quantization, is 255 for image samples, so this is just 
(val - valsecond) / (2 * 255)
 
Which is (basically) what we calculate.

Now, for the question about the (idxsecond <= idx) term: This is discussed in Appendix B ("Experimental Details") of the paper. To quote:

"In DSSN, following Levine & Feizi (2020a) (discussed in Section 1.1), if two classes tie in the number of “votes”, we predict the first class lexicographically: this means that we can certify robustness up to _and including_ the radius ρ, because we are guaranteed consistent behavior in the case of ties. Reported certified radii for DSSN should therefore be interpreted to guarantee robustness even in the ||x − x′||_1 = ρ case. (This is not a meaningful distinction in randomized methods where the space is taken as continuous)."

You can see our previous paper that was referred to, which uses a similar mechanism, here: https://openreview.net/forum?id=YUGG2tFuPM.

To explain briefly, note that the explanation about the certified radius computation I gave above applies only for certifying that class A will be returned  _inside_ of the ball where ||x − x′|| < d/(2c). However, when we report a certificate, we would like to certify a radius for which the returned classification will remain consistent *up to and including* radius ρ. This can knock a small amount off of our certificate. 

Very concretely, note that for every 1/q unit of displacement of x' away from x, 1/(2 * lambda)-Lipshitzness implies that "val" (= p_A * q * 2 * lambda) can decrease by at most 1 "vote", while "valsecond"  (= p_B * q * 2 * lambda) can increase by at most  1 "vote". 

**If "val - valsecond" is _odd_**, then we can certify to a radius of 

rho := (1/q) *(val - valsecond - 1 )/2   == (1/q) * Floor((val - valsecond  )/2)  == (1/q) * Floor((val - valsecond - 1 )/2).

 At this radius, p_A will still be above p_B  at distance rho, because p_A can decrease by at most (val - valsecond - 1 )/2 "votes" and p_B can increase by at most  (val - valsecond - 1 )/2  "votes", so the "gap" will close by at most (val - valsecond - 1 ), and so still will be at least 1. Note that the (idxsecond <= idx) term is irrelevant in this case, because its value is masked by the floor function (implemented as .astype(int))

**However, if "val - valsecond" is _even_**, then, at a radius of exactly
(1/q) *(val - valsecond )/2  == (1/q) *Floor((val - valsecond )/2)
it is possible that p_A and p_B will exactly **tie**. This is because the gap is initially (val - valsecond) so if p_A decreases by (val - valsecond )/2 and p_B increases by this amount, then p_A = p_B at radius rho.

In this case, the behavior of the smoothed classifier in the case of "tie votes" becomes relevant. If we can guarantee that the smoothed classifier will return class A instead of class B in the case that the votes tie, then we can certify to a radius of
(1/q) *(val - valsecond )/2 == (1/q) *Floor((val - valsecond )/2).
However, if we can't guarantee this, then we can only certify to a radius of
(1/q) *((val - valsecond )/2 -1) == (1/q) *Floor((val - valsecond -1 )/2),
where P_A will still exceed P_B by 2 "votes". 

Therefore, in order to be able to return slightly larger certificates, we set a deterministic behavior for the smoothed classifier in the case of "tie votes": in this case, we always return the lower-index classification. This means that we can check whether class A or class B would be returned in the case of a tie, and decide to certify a slightly larger radius if class A would be preferred. This is why our certificate has the form, essentially

(1/q) *Floor((val - valsecond - (prefer class B in case of tie?)  )/2),

or, as in the code:

 ((val - valsecond - (idxsecond <= idx))/2).astype(int)/255.

Note that this is the reason to use "kind='stable'" when argsorting the indices on line 2 of the method. This guarantees that the actual returned classification, the variable 'idx', is the **lowest-indexed** class with the highest votes number of votes, and also that idxsecond is the **lowest index** of any class that has the runner-up number of votes. Note that any other class C either has a higher index but the same number of votes, or a smaller number of votes by at least one, so the value of "val - val_C - (prefer class C in case of tie?)" would be at least as large as "val - valsecond - (prefer class B in case of tie?)".

Let me know if this explanation is helpful.

**Comment by Rui-Wang-813** [comment]
Thank you very much. It is so kind of you to give such a detailed explanation! I understand it now. : ) I did not consider that much in empirical (odd/even, floor) aspect. 

Just out of curiosity, I read another [paper](https://arxiv.org/abs/2210.01787) implying an approach to calculate certified radius given Lipschitz constant:
![image](https://github.com/user-attachments/assets/f4439ee0-ff68-4322-906f-fc50ddc1c3de)
Do you think this is applicable (or tighter) in your work if using margins between logits instead of each single logit?

**Comment by alevine0** [comment]
The paper you linked is using a slightly different definition than we do of what a "Lipschitz classifier" is. In particular, in our paper, we only prove that each individual logit (e.g., p_A(x)), as a **scalar value**, is 1/(2 * lambda)-Lipschitz with respect to the L_1 norm; that is:
|p_A(x) - p_A(x')| <= 1/(2 * lambda) * ||x-x'||_1
By contrast, the paper you linked requires that the entire output of the classifier f, as a **vector**, is Lipschitz with respect to some  L_p norm
<img width="1009" alt="image" src="https://github.com/user-attachments/assets/b0c7955f-5590-45b6-a11b-7c6fedffe521">
This is a more-strict definition of Lipschitzness than the one we use. In fact, if we just look at the binary classification case, because both p_A and p_B must change by the same magnitude (because p_A= 1 -p_B), we have that our method would guarantee a lipschitz constant twice as large under their definition as under ours. Then, using a Lipschitz constant of 1/lambda, and plugging into their Proposition 3.1, we get a certificate of rho = lambda * sqrt(2) *d /2, which is smaller than what we get using the certification formula from our paper: rho = 2* lambda * d / 2 == lambda * d. So this doesn't give a tighter bound for our particular method.

**Comment by Rui-Wang-813** [comment]
Thank you very much for your detailed and patient reply!
